internal/dispatch/remote/cluster.go

```diff
@@ -112,28 +146,35 @@ func (cr *clusterDispatcher) DispatchReachableResources(
 		return err
 	}
 
-	client, err := cr.clusterClient.DispatchReachableResources(ctx, req)
+	withTimeout, cancelFn := context.WithTimeout(ctx, cr.dispatchOverallTimeout)
+	defer cancelFn()
+
+	client, err := cr.clusterClient.DispatchReachableResources(withTimeout, req)
 	if err != nil {
 		return err
 	}
 
 	for {
-		result, err := client.Recv()
-		if errors.Is(err, io.EOF) {
-			break
-		}
-
-		if err != nil {
-			return err
-		}
-
-		serr := stream.Publish(result)
-		if serr != nil {
-			return serr
+		select {
+		case <-withTimeout.Done():
+			return withTimeout.Err()
+
+		default:
+			result, err := client.Recv()
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+
+			if err != nil {
+				return err
+			}
```

Comment:
shouldn't this just be another `case`, instead of error handling?

Reply:
EOF is the err returned from `client.Recv()`, so it can't be another case

Reply:
> nit:
> 
> ```go
> if errors.Is(err, io.EOF) {
>     return nil
> } else if err != nil {
>     return err
> }
> ```

I actually dislike this style, but changed because it is slightly more efficient

Reply:
yeah it's a little ugly, but i prefer it because it creates one shared logical scope